### ISSUE TITLE
Some things I feel need fixing/adding

### ISSUE THREAD

**Issue by drewauff** [walles/px#88]
As per title:
1) The ability to change mode to CPU Usage (why that isn't there is beyond me)

2) After search/filter and killing a process, add option to automaticly return to process list.

3) An option to hide the headers

4) The ability to pipe through fzf for super quick process killing (px)

5) An option to hide/remove columns

6) Performance increase. There is a lack of "snappiness"

I guess thats all for now, I will update this post if I come up with anything else.

**Comment by walles** [comment]
Hello @drewauff, thanks for reporting!

Regarding 1, CPU sort, is this for `px` or `ptop`?

Regarding 2, returning to process list, I just tried doing `grep foo`, then started `ptop` in another terminal, searched for `grep` and killed it in `ptop`. This takes me back to the `ptop` main screen. What is is that you want changed? Clearing the filter string in this case?

Regarding 3, 4 and 5, are these all part of streamlined process killing? I really want this, but I haven't been able to decide how exactly I want it to work. Basically I'd like to type one command (`pxkill`?) with an optional filter string, possibly do more filtering and select a process (like `fzf`, possibly using `fzf`), then have it killed.

Regarding 6, I agree on the performance. I know what **I** think feels slow, but what do **you** feel is slow? `px` or `ptop`? Remote or local? Startup, general screen refreshes or is it when you do some particular operation?

**Comment by drewauff** [comment]
Sort for both.

yes, I think clearing the filers would do it. 

Yes, its as simple as setting up a zle widget.

You run the the widget with whatever shortcut(s) you want, the list pops up, start typing and the list narrows down till you find what your after. Its Very bloody fast.

my ps widget code :

`fzf-ps-widget() {
	  ps acxh -eo  %cpu,%mem,pid,command -r≥w | sort -rnk 1 | head -n20 | fzf --no-tac --preview-window=hidden | awk '{print $3}' | xargs kill -9
}
zle -N fzf-ps-widget
bindkey '^k' "fzf-ps-widget"
`
pkill can be used, I had thought of using killall or pkill -a but I stuck with ps mostly because it was the first one to work properly. Should not that px will have to be used with fzf.

Both px and ptop are a bit sluggish to load, ptop just feels a bit "laggy" in terms of operation.

As for the columns, All that I need is PID, CPU% MEM%, short Command (no need for path, makes for cleaner ui), maybe adding option to hide the headers in ptop.

**Comment by drewauff** [comment]
here is a "snapshot" of px piped through fzf :

![Screen Shot 2021-08-16 at 12 41 27 pm](https://user-images.githubusercontent.com/13850060/129504423-387572be-7f55-4abe-90cf-7b6fb6bd5468.png)

I used `px | awk '{print $1,$2,$4,$6}' |column -t |  fzf --tac`

note: cant remove large space after "command" , awk recognises it as part of $2 and cut just totally borks the display.

**Comment by walles** [comment]
For 6, lack of `ptop` snappiness, I did some work in 1.5.2: <https://github.com/walles/px/releases/tag/1.5.2>

Compared to 1.5.0, if you start `ptop`, then press and hold down arrow, for me:
* 1.5.0 slides the selection down in a choppy motion
* 1.5.2 slides the selection down in a smoother motion

The motion still isn't always perfect, but at least when I try this it's a good improvement.

**Comment by drewauff** [comment]
1.5.2 is smoother , not much "snappier" also after exiting the ptop is still visible unlike most (99%) of other programs.

**Comment by walles** [comment]
Thanks for the `fzf` examples, I only used its [`fish`](https://github.com/fish-shell/fish-shell) integration so far!

Anyway, I experimented a bit with `fzf` and your snippets, and this gets you a `kill` command where you can pick processes with no `px` modifications necessary:

```
px | fzf --height=20 --no-hscroll --tac --no-sort --header-lines=1 | awk '{print $1}' | xargs kill -9
```

When I try the above I get:
* the most interesting process highlighted by default (`px` sort order [documented here](https://github.com/walles/px#px), a process is "interesting" if it has used lots of CPU, lots of memory or been started recently)
* correct headers straight from `px`
* all the standard `fzf` behaviour with fuzzy matching etc

**Comment by drewauff** [comment]
Interesting (see what I did there :) ) idea, but being able to sort by cpu% usage  is a must. That way at a glance you can see whats bogging down your system.

Also, just had another idea, is there a way to for "crashed" apps to the the top of the list?  ie. sometimes an app crashes but its using very little resources but you'd never know that its crashed until you check your processes and because its using very little resources you have to "search" for it, would it not be much quick if you hit the widget/command hotkey and straight away you are presented with the crashed process (in red obviously) at the top of the list.

Anyway, just a (good) thought.

**Comment by walles** [comment]
Let's say you have two processes:
1. One process using 90% of your RAM but no CPU.
2. One process using almost no RAM but 1% CPU.

In this scenario, the first process' memory usage will make other processes slower by:
1. Eating up memory that would otherwise be used as disk cache for other apps
2. Likely causing swapping, possibly affecting other apps by replacing fast memory accesses with less fast IO

So to me, the memory hogging process does contribute to bogging down my system, and I want `px` to show me this process among the interesting ones.

In your opinion, which process would you consider bogging down your system the most?
1. The one using almost all your RAM but no CPU?
2. The one using almost no RAM and almost no CPU?

**Comment by drewauff** [comment]
3. The one using almost all RAM and CPU.

To explain Further,

When an app crashes, more often than not it spikes the cpu to 100%+. spiked cpu% is the first sign of a crashing/crashed app. In most activity/process monitors, processes either crashed or spiking the cpu% are show in red accompanied by an error message of some description.

In my opinion, its best to handle this via flags while incorporating a configurable shortcut/hotkey to switch between whichever mode you want ie ctrl-1 = cpu%  ctrl-2 mem% etc. maybe a custom sort option in there as well.

As I said previously I think its also very important to pin crashed or spiking programs to the top of the list.

**Comment by drewauff** [comment]
Problems detected, please send this text to one of:
* https://github.com/walles/px/issues/new
* johan.walles@gmail.com

px version: 1.5.3

Python version: 3.9.6 (default, Jun 29 2021, 06:20:32) 
[Clang 12.0.0 (clang-1200.0.32.29)]

Platform info: macOS-10.15.7-x86_64-i386-64bit

2021-08-24 10:24:32AEST ERROR    Uncaught Exception
Traceback (most recent call last):
  File "/usr/local/Cellar/px/1.5.3/libexec/lib/python3.9/site-packages/px/px.py", line 98, in main
    _main(argv)
  File "/usr/local/Cellar/px/1.5.3/libexec/lib/python3.9/site-packages/px/px.py", line 248, in _main
    procs = sorted(procs, key=operator.attrgetter("cpu_percent"))
TypeError: '<' not supported between instances of 'float' and 'NoneType'

**Comment by walles** [comment]
Oops, 1.5.4 released with a fix in it. Test added.

I happened to test it in Python 2 only (where it worked), but not in Python 3 (where it didn't).

**Comment by drewauff** [comment]
works..does not like `head` but `tail` works.

Anyway to remove/hide the user/group?

Im also trying to find a way to get fzf to live update the list..so far not so good.

Another thing do you use ansi color codes for the color? fzf has a switch to enable ansi coloring

**Comment by drewauff** [comment]
just found this and it works (kinda)
`--bind 'ctrl-r:reload(ps -ef)' --header 'Press CTRL-R to reload'`
If I could get to self update well...

**Comment by walles** [comment]
Both `px` and `ptop` use ANSI codes to highlight parts of their output.

I think I kind of know the answer to this one, but still, if you want a live updating list, why don't you just use `ptop`?

`ptop` lists processes by CPU usage, it live updates and you can pick processes for killing.

**Comment by drewauff** [comment]
hmmm ok thats good I think..

I have tried ptop, but px via much faster:

Tap shortcut key(s) ==> starting type process to be killed ==> hit enter to kill 

process is killed and you are returned to prompt.

Just thought it was a good idea if Im honest.

**Comment by walles** [comment]
I think it would be hard to do much better than binding this to your shortcut key:

```
px --sort=cpupercent | fzf --bind 'ctrl-r:reload(px --sort=cpupercent)' --height=20 --no-hscroll --tac --no-sort --header-lines=1 | awk '{print $1}' | xargs kill -9
```

This list can be reloaded with <kbd>ctrl</kbd><kbd>r</kbd>.

With a self updating list (like every second or so), you risk killing the wrong process if you hit <kbd>return</kbd> just as the list changes, so that could be tricky to get right UI wise.

Regarding ANSI coloring, `px` does that when printing to a terminal, but not when piped. This could get its own switch of course, but right now it's not there.

And as a side note, adding `--preview='px {1}'` to the list will give you a preview for the currently selected process. I don't really think I like it, but you should try it and decide for yourself!

**Comment by drewauff** [comment]
👍🏽 ...can you add flags to hide columns ie. groups? I can remove the "groups" column using awk but it destroys the output however that can be somewhat sorted using `columns -t` this results in the output being "stretched" ie. the space after a column in removed is huge.

Other than that bloody awesome! 👍🏽

**Comment by drewauff** [comment]
I should not that the reason for the "huge" gap is due to long process names, maybe theres a way to truncate them somehow, to a user defined degree?

**Comment by walles** [comment]
Can you send me a screenshot of what my suggested `fzf` pipeline ^ looks with your preferred terminal window size / fonts / whatever? I'd like to see how much screen real estate you have and what information is visible for you. Sounds like you have fewer characters available per line than I want to know what that looks like.

Also, are these the columns you care about?
* `PID`
* `COMMAND`
* Not `USERNAME`
* `CPU`
* `CPUTIME`
* `RAM`
* Not `COMMANDLINE`

Here's what it looks like for me, I scrolled up a lot to find the long command name culprit (`CoreEmbeddedSpeechRecognition/embeddedspeech` in my case):
<img width="843" alt="gap-explanation" src="https://user-images.githubusercontent.com/158201/137614764-c1d93125-a497-47aa-b361-ab253cb12f0b.png">

**Comment by drewauff** [comment]
yes thats exactly what I have and I thought that truncation would be best (line wrap would just be silly).

Other than needing truncation, after the username the gap is also quite big.

using `columns -t` and/or `tr -s ' '` , does alright except it stretches the output to the full width of the terminal window which creates even bigger gaps also depends on the use of awk or cut. Both yield not so nice results.


![Screen Shot 2021-10-17 at 7 27 10 pm](https://user-images.githubusercontent.com/13850060/137618622-c16e8ae9-0b73-4d04-bac1-dee963a948a4.png)

**Comment by walles** [comment]
So if you got a `--truncate` option, that would solve your problem?

It would ensure everything except `COMMANDLINE` fits within the terminal window width, by truncating usernames / commands if necessary.

`COMMANDLINE` would be used as filler if there's room.

**Comment by walles** [comment]
In your example, all columns are visible.

Why is the spacing a problem for you?

Is it because:
1. You are losing information?
2. The rows are ugly / hard to read?
3. ... something else, what?

**Comment by drewauff** [comment]
Spacing is the issue when removing username column regardless of method.

If you look at the screenshot I posted, the process name is right against the username and the spacing is very wide on the other side of username.

If there a was a way to remove the username without the use of awk or cut then we could see what spacing need to be adjusted.

So we need to find a way to remove the username programatically then go from there.

**Comment by walles** [comment]
Why is removing the username column important to you? How is having the username there a problem?

If the spacing was fixed with the username still in place, would that solve your problem?

**Comment by drewauff** [comment]
this screenshot wil make it more clear, also I know which user is on this machine..me and root thats it. so wasted space wich could be better utilised.

You can see how its cut off..without that column and a little of the space around it, more info can be shown, especially with the preview open

![Screen Shot 2021-10-17 at 10 45 20 pm](https://user-images.githubusercontent.com/13850060/137625999-9179cfbb-ef1a-40c3-b3e2-bedb7ee51d77.png)

**Comment by walles** [comment]
Thanks, this screenshot does clarify things!

Given this screenshot, just removing the user name column would make sense.

And note to self really, given this use case it's impossible for `px` to autodetect the available width.

So I'd say just having a `--no-username` switch would resolve this.

I know this is kind of what you've been asking for all along, but I like to understand what problem I'm solving before changing stuff.

**Comment by drewauff** [comment]
just did a little bit of testing, cannot remove the username column without breaking the preview...

I tried with awk and cut, then tried using column to "reorganise" it columns the remove extra whitespace with tr.. no breaks the preview..

**Comment by walles** [comment]
Just released this version with a `--no-username` switch: <https://github.com/walles/px/releases/tag/1.5.5>

If you're using Homebrew to install `px`, here's the corresponding Homebrew bumping PR: <https://github.com/Homebrew/homebrew-core/pull/87458>

Suggested command line with previews and no user names, stolen from the updated `README`:

```
px --sort=cpupercent --no-username | fzf --preview='px --color {1}' --bind 'ctrl-r:reload(px --sort=cpupercent --no-username)' --height=20 --no-hscroll --tac --no-sort --header-lines=1 | awk '{print $1}' | xargs kill -9
```

**Comment by drewauff** [comment]
Yes, thats how ive been updating...

Ive been trying to get make the code easier to deal with, resulting in this
![Screen Shot 2021-10-18 at 10 00 30 am](https://user-images.githubusercontent.com/13850060/137648087-755f5968-8685-476d-8d86-97ed1f9bd9f7.png)

except it errors out the pipe..any ideas?

**Comment by drewauff** [comment]
I made the code a little easier to read:
except Im getting a pipe error..any ideas?
![Screen Shot 2021-10-18 at 10 00 30 am](https://user-images.githubusercontent.com/13850060/137648829-a27d1d35-0ddf-4451-8d54-2e4142785f7e.png)

**Comment by drewauff** [comment]
its ok figured it out..pipe has to go on the same line as `--header-lines` after the `1`

**Comment by walles** [comment]
> its ok figured it out..pipe has to go on the same line as `--header-lines` after the `1`

Or you could add a `\` after the `--header-lines=1` to escape the linebreak, that should work too.

**Comment by walles** [comment]
If this works for you I'll be closing this ticket now!

**Comment by drewauff** [comment]
Ok, will do..

Thank you for so much of your time and patience.

I appreciate you

Thank you 👍🏽 

-
Andy

**Comment by walles** [comment]
Thank *you* for your persistence Andy :)

**Comment by walles** [comment]
The CPU sorting may or may not be needed as much after 723057b1ac84abd289d4901b71857e7d03dba331, I think that should improve the default sort order.

Released as 3.4.0 in case you want to try it out!
